src/lib.rs

```diff
@@ -77,299 +63,138 @@
-
-    #[error("Invalid from publish API: can't find published message id")]
-    ResponseLacksId,
+    MessageRoute(Uuid),
 
+    /// Could not parse a schema URL
     #[error("Could not parse `{1}` as a schema URL")]
-    SchemaUrlParsing(#[source] url::ParseError, String),
+    SchemaUrlParse(#[source] url::ParseError, String),
 
+    /// Could not resolve the schema URL
     #[error("Could not resolve `{0}` to a schema")]
-    SchemaUrlResolving(url::Url),
+    SchemaUrlResolve(url::Url),
 
+    /// Could not validate message data
     #[error("Message data does not validate per the schema: {0}")]
     DataValidation(String),
-}
 
-/// Message publishers.
-///
-/// This is used to interface with arbitrary pub-sub services.
-pub trait Publisher {
-    /// The list of identifiers for successfully published messages.
-    type MessageIds;
+    /// Publisher failed to publish a message
+    #[error("Publisher failed to publish a message batch")]
+    Publisher(#[source] Box<dyn std::error::Error + Send + Sync>),
 
-    /// Publish a Hedwig message.
-    ///
-    /// The Vector of topic-message pairs is passed in. This vector is not sorted and ordering has
-    /// no semantic meaning.
-    ///
-    /// Shall return `Ok` only if all of the messages are successfully published.
-    fn publish(
-        &self,
-        messages: Vec<(&'static str, ValidatedMessage)>,
-    ) -> Result<Self::MessageIds, PublishError>;
+    /// Publisher failed to publish multiple batches of messages
+    #[error("Publisher failed to publish multiple batches (total of {} errors)", _1.len() + 1)]
+    PublisherMultiple(
+        #[source] Box<dyn std::error::Error + Send + Sync>,
+        Vec<Box<dyn std::error::Error + Send + Sync>>,
+    ),
 }
 
-/// A publisher that uses Google PubSub. To use this class, add feature `google`.
-///
-/// # Examples
-///
-/// ```no_run
-/// # #[cfg(feature = "google")]
-/// # use hedwig::{Publisher, GooglePublisher};
-/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
-/// # #[cfg(feature = "google")]
-/// let publisher = GooglePublisher::new(String::from("/home/.google-key.json"), "myproject".into())?;
-/// # Ok(())
-/// # }
-/// ```
-#[cfg(feature = "google")]
-#[allow(missing_debug_implementations)]
-pub struct GooglePublisher {
-    client: Pubsub<Client, oauth2::ServiceAccountAccess<Client>>,
-    google_cloud_project: String,
-}
-
-#[cfg(feature = "google")]
-impl GooglePublisher {
-    /// Create a new instance of GooglePublisher
-    /// For now, this requires explicitly passing in the path to your credentials file, as well as the name of the
-    /// project where pubsub infra lives.
+/// The special result type for [`Publisher::publish`](trait.Publisher.html)
+#[derive(Debug)]
+pub enum PublisherResult<Id> {
+    /// Publisher succeeded.
     ///
-    /// # Arguments
+    /// Contains a vector of published message IDs.
+    Success(Vec<Id>),
+    /// Publisher failed to publish any of the messages.
+    OneError(
```

Comment:
Dunno if the separation of all/some is useful - you presumably want to know about failures in either case.

Reply:
I think my thinking back then was that it at very least avoids an allocation (of the `Vec`), but I might be wrong and there might have been some deeper underlying plan.

Reply:
Ehh.. seems like a non-issue for us? We'll always publish multiple right..

Reply:
Sounds accurate.

This perhaps matters more if the `Hedwig::publish` API is used by somebody to publish single messages. Provided the underlying `Publisher` instance utilizes `OneError` to signal errors, that could end up being a significant improvement, though I guess its also an instance of premature optimization leaking out.